From 0ded438a6ab7e8d11922582bf90a7dfa6b7fdcd4 Mon Sep 17 00:00:00 2001
From: Cyan <cyan@cyano.uk>
Date: Mon, 10 Jul 2023 00:14:13 +0800
Subject: [PATCH 17/25] commands: add new command 'pause'

- Simple enough, this command prints out a prompt, either pre-defined or
  user specified, and awaits a key stroke from the user.
---
 grub-core/Makefile.core.def |  5 ++++
 grub-core/commands/pause.c  | 54 +++++++++++++++++++++++++++++++++++++
 2 files changed, 59 insertions(+)
 create mode 100644 grub-core/commands/pause.c

diff --git a/grub-core/Makefile.core.def b/grub-core/Makefile.core.def
index b3b847766..542262731 100644
--- a/grub-core/Makefile.core.def
+++ b/grub-core/Makefile.core.def
@@ -1058,6 +1058,11 @@ module = {
   common = commands/password_pbkdf2.c;
 };
 
+module = {
+  name = pause;
+  common = commands/pause.c;
+};
+
 module = {
   name = play;
   x86 = commands/i386/pc/play.c;
diff --git a/grub-core/commands/pause.c b/grub-core/commands/pause.c
new file mode 100644
index 000000000..39d193a17
--- /dev/null
+++ b/grub-core/commands/pause.c
@@ -0,0 +1,54 @@
+/* pause.c - Wait for a keystroke with or without custom messages  */
+/*
+ *  GRUB  --  GRand Unified Bootloader
+ *  Copyright (C) 2006,2007,2010  Free Software Foundation, Inc.
+ *
+ *  GRUB is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  GRUB is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with GRUB.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <grub/dl.h>
+#include <grub/misc.h>
+#include <grub/extcmd.h>
+#include <grub/err.h>
+#include <grub/i18n.h>
+#include <grub/term.h>
+
+GRUB_MOD_LICENSE ("GPLv3+");
+
+static grub_err_t
+grub_cmd_pause (grub_extcmd_context_t ctxt, int argc, char **args)
+{
+  struct grub_arg_list *state = ctxt->state;
+  if (argc != 1) {
+    grub_printf ("%s", N_("Press any key to continue..."));
+  } else {
+    grub_printf ("%s", args[0]);
+  }
+  (void) grub_getkey ();
+  return GRUB_ERR_NONE;
+}
+
+static grub_extcmd_t cmd;
+
+GRUB_MOD_INIT(pause)
+{
+  cmd = grub_register_extcmd ("pause", grub_cmd_pause, 0,
+			      "STRING", N_("Wait for a key stroke with a message."),
+			      0);
+}
+
+GRUB_MOD_FINI(pause)
+{
+  grub_unregister_extcmd (cmd);
+}
-- 
2.43.0

